Update of realizations diffLin2Lin and diffExp2Lin of 45_carbonprice …

laurinks opened the pull request:
…(linear curve going through (2020,25) instead of (2010,0), using new switch cm_co2_tax_startyear instead of cm_co2_tax_2020); rename switch c_peakBudgYr to cm_peakBudgYr due to usage in diffLin2Lin of 45carbonprice

## Purpose of this PR

Update diffLin2Lin realization of 45_carbonprice to ensure more near-term realism in carbon pricing by making the linear curve go through the point (2020,25) instead of (2010,0) for developed regions. The choice of the point (2020,25) is based on 25$/t CO2eq being the historical CO2 price for EUR, which is the highest among the regions.

## Type of change

- [ ] Fundamental change

## Checklist:

- [ ] My code follows the [coding etiquette](https://github.com/remindmodel/remind/blob/develop/main.gms#L80)
- [ ] I performed a self-review of my own code
- [ ] I explained my changes within the PR, particularly in hard-to-understand areas
- [ ] I checked that the [in-code documentation](https://github.com/remindmodel/remind/blob/develop/main.gms#L120) is up-to-date
- [ ] I adjusted the reporting in [`remind2`](https://github.com/pik-piam/remind2) where it was needed
- [ ] I adjusted `forbiddenColumnNames` in [readCheckScenarioConfig.R](https://github.com/remindmodel/remind/blob/develop/scripts/start/readCheckScenarioConfig.R) in case the PR leads to deprecated switches
- [ ] All automated model tests pass (`FAIL 0` in the output of `make test`)
- [ ] The changelog `CHANGELOG.md` [has been updated correctly](https://gitlab.pik-potsdam.de/rse/rsewiki/-/wikis/Standards-for-Writing-a-Changelog)

## Further information (optional):

* Test runs are here: 
* Comparison of results (what changes by this PR?):

laurinks commented on modules/45_carbonprice/diffExp2Lin/datainput.gms:
```diff
@@ -27,18 +26,18 @@ if(cm_co2_tax_spread eq 10,
+p45_phasein_2025ratio(regi)$(p45_gdppcap2015_PPP(regi) gt 10  and p45_gdppcap2015_PPP(regi) le 15) = 0.4; !! CHA, LAM, MEA
+p45_phasein_2025ratio(regi)$(p45_gdppcap2015_PPP(regi) gt 15  and p45_gdppcap2015_PPP(regi) le 20) = 0.6; !! REF
+p45_phasein_2025ratio(regi)$(p45_gdppcap2015_PPP(regi) gt 20) = 1; !! EUR, JPN, USA, CAZ, NEU
 );
```
This is just an updating of the comments which had not been correct before.

laurinks commented on main.gms:
```diff
@@ -509,6 +509,15 @@ parameter
+*' * (-1): default setting equivalent to no carbon tax
+*' * (any number >= 0): co2 tax in start year [if cm_iterative_target_adj eq 0]; 
+*' *                    exogenous initialization of co2 tax in start year [if cm_iterative_target_adj ne 0]
+*'
```
Introducing new switch cm_co2_tax_startyear to avoid using cm_co2_tax_2020 in new diffLin2Lin and diffExp2Lin implementations. Why? 2020 is a historical year by now, and it is not intuitive to adjust carbon pricing from 2025 or 2030 via switch for year 2020.

laurinks commented on modules/45_carbonprice/diffLin2Lin/datainput.gms:
```diff
@@ -27,36 +27,38 @@ if(cm_co2_tax_spread eq 10,
+                                + (s45_co2_tax_startyear - s45_co2_tax_2020) / (cm_startyear - 2020) !! Yearly increase of CO2 price 
+                                  * (t.val - 2020) ;
+p45_CO2priceTrajDeveloped(t)$(t.val gt 2110) = p45_CO2priceTrajDeveloped("2110"); !! set taxes constant after 2110
+*** for peak budget runs (if cm_iterative_target_adj = 6|7|9), the adjustment of the CO2 price trajectory after the peak year is made in core/postsolve.gms
```
This is formula is the core of the change. It also appears in the postsolve of this module.

laurinks commented on modules/45_carbonprice/diffLin2Lin/postsolve.gms:
```diff
@@ -9,15 +9,31 @@
+                                    + (pm_taxCO2eq("2110",regi) - s45_co2_tax_2020) / (2110 - 2020) !! Yearly increase of CO2 price that interpolates between s45_co2_tax_2020 in 2020 and pm_taxCO2eq in 2110
+                                      * (t.val - 2020) ;
+      p45_CO2priceTrajDeveloped(t)$(t.val gt 2110) = pm_taxCO2eq(t,regi);
+    );
```
REMIND adjusts the CO2 price trajectory in `core/postsolve.gms` between iterations in order to achieve the prescribed budget. This adjustments happens by multiplying the entire CO2 price trajectory with the factor `p_factorRescale_taxCO2_Funneled`. After this multiplication, the new CO2 price trajectory no longer goes through the starting point (2020,25) and thus we need to adapt it here. To this end, we re-create the linear curve through the points (2020,25) and (peak year, adjusted CO2 price in peak year) (resp. through (2110, adjusted CO2 price in 2110)).

laurinks replied:
Note that this was not needed in previous version which used (2010,0) as the starting point because this was stable under multiplication.

LaviniaBaumstark commented on modules/45_carbonprice/diffExp2Lin/datainput.gms:
```diff
@@ -47,15 +46,16 @@ display p45_phasein_2025ratio;
+  abort "please choose a valid cm_co2_tax_startyear"
+elseif cm_co2_tax_startyear gt 0,
 *** convert tax value from $/t CO2eq to T$/GtC
-p45_CO2priceTrajDeveloped("2020") = cm_co2_tax_2020 * sm_DptCO2_2_TDpGtC;
```
do we still need cm_co2_tax_2020?

laurinks replied:
Yes, it is used in the realizations `diffCurvPhaseIn2Lin`, `expoLinear`, `exponential`, and `linear`. I plan to delete all these realizations except  `expoLinear` because there functionalities are covered by `diffExp2Lin `and `diffLin2Lin`. Then I can also delete this switch. But I would like to wait with the clean up until my next PR (which will have further developments on diffLin2Lin and will come in August after my vacation).

LaviniaBaumstark replied:
I agree and thanks for cleaning up

orichters commented:
Can you please run `comparescenconf config/scenario_config.csv` in your REMIND folder and paste the human-readable diff here? Thanks.

laurinks commented:
File comparison: /p/projects/rd3mod/github/repos/remindmodel/remind/develop/config/scenario_config.csv -> config/scenario_config.csv
Columns deleted: c_peakBudgYr
Columns added:   cm_co2_tax_startyear, cm_peakBudgYr
Renamed columns: -
Renamed rows:    -

Changes in the scenarios:
~ SSP2-PkBudg500:
    carbonprice: diffCurvPhaseIn2Lin -> diffLin2Lin (default: none)
    cm_co2_tax_2020: 100 -> NA (default: -1)
    cm_co2_tax_startyear: NA -> 200
    cm_peakBudgYr: NA -> 2045
~ SSP2-PkBudg650:
    carbonprice: diffCurvPhaseIn2Lin -> diffLin2Lin (default: none)
    cm_co2_tax_2020: 100 -> NA (default: -1)
    cm_co2_tax_startyear: NA -> 75
    cm_peakBudgYr: NA -> 2055
~ SSP2-PkBudg1050:
    carbonprice: diffCurvPhaseIn2Lin -> diffLin2Lin (default: none)
    cm_co2_tax_2020: 60 -> NA (default: -1)
    cm_co2_tax_startyear: NA -> 40
    cm_peakBudgYr: NA -> 2100
~ SSP2-EU21-PkBudg500:
    cm_peakBudgYr: NA -> 2080
~ SSP2-EU21-PkBudg650:
    cm_peakBudgYr: NA -> 2080
~ SSP2-EU21-PkBudg1050:
    cm_peakBudgYr: NA -> 2100
~ SSP1-PkBudg650:
    cm_peakBudgYr: NA -> 2080
~ SSP1-PkBudg1050:
    cm_peakBudgYr: NA -> 2100
~ SSP5-PkBudg650:
    cm_peakBudgYr: NA -> 2080
~ SSP5-PkBudg1050:
    cm_peakBudgYr: NA -> 2100
~ SDP_MC-PkBudg650:
    cm_peakBudgYr: NA -> 2045
~ SDP_EI-PkBudg650:
    cm_peakBudgYr: NA -> 2045
~ SDP_RC-PkBudg650:
    cm_peakBudgYr: NA -> 2045
~ SSP2_lowEn-PkBudg650:
    cm_peakBudgYr: NA -> 2080
~ SSP2_lowEn-PkBudg1050:
    cm_peakBudgYr: NA -> 2100

$allwarnings
NULL

orichters commented on main.gms:
```diff
@@ -379,7 +379,7 @@ $setglobal emicapregi  none           !! def = none
 *' * (diffCurvPhaseIn2Lin): [REMIND 2.1 default for validation peakBudget runs, in combination with iterative_target_adj = 9] curved convergence of CO2 prices between regions until cm_CO2priceRegConvEndYr; developed countries have linear path from 0 in 2010 through cm_co2_tax_2020 in 2020;
 *' * (diffExp2Lin): quadratic convergence of CO2 prices between regions until cm_CO2priceRegConvEndYr; starting level of differentiation in cm_startyear is given by cm_co2_tax_spread; developed countries have exponential path until peak year (with iterative_target_adj = 9) or until 2100 (with iterative_target_adj = 5);
-*' * (diffLin2Lin): [Variant of diffCurvPhaseIn2Lin, will be further developed before becoming the new default] quadratic convergence of CO2 prices between regions until cm_CO2priceRegConvEndYr; starting level of differentiation in cm_startyear is given by cm_co2_tax_spread; developed countries have linear path until peak year (with iterative_target_adj = 9) or until 2100 (with iterative_target_a ...
+*' * (diffLin2Lin): [Variant of diffCurvPhaseIn2Lin, will become the new default] quadratic convergence of CO2 prices between regions until cm_CO2priceRegConvEndYr; starting level of differentiation in cm_startyear is given by cm_co2_tax_spread; developed countries have linear path until peak year (with iterative_target_adj = 9) or until 2100 (with iterative_target_adj = 5);
```
```suggestion
*' * (diffLin2Lin): [default for peak budget runs, derived from diffCurvPhaseIn2Lin] quadratic convergence of CO2 prices between regions until cm_CO2priceRegConvEndYr; starting level of differentiation in cm_startyear is given by cm_co2_tax_spread; developed countries have linear path until peak year (with iterative_target_adj = 9) or until 2100 (with iterative_target_adj = 5);
```

laurinks replied:
Thanks! I raised also the question [here](https://github.com/remindmodel/development_issues/issues/295#issuecomment-2236814015) when the good timing would be for making the change in all the other defaults.

orichters replied:
I would adjust it in all scenarios of `scenario_config.csv` simultaneously.

laurinks replied:
Done.

orichters reviewed (approved):
Approve, but I would prefer to see all scenarios in scenario_config.csv being adapted. Then just wait for the next round of AMTs and check the compareScenarios files.
